suprb/rule/matching.py

```diff
@@ -78,3 +79,162 @@ def min_range(self, min_range: float):
+        return np.prod(diff)
+
+    def copy(self):
+        return CentreSpread(self.bounds.copy())
+
+    def _validate_bounds(self, X: np.ndarray):
+        """Validates that bounds have the correct shape."""
+
+        if self.bounds.shape[1] != 2:
+            raise ValueError(f"specified bounds are not of shape (-1, 2), but {self.bounds.shape}")
+
+        if self.bounds.shape[0] != X.shape[1]:
+            raise ValueError(f"bounds- and input data dimension mismatch: {self.bounds.shape[0]} != {X.shape[1]}")
+
+    def clip(self, bounds: np.ndarray):
+        self.bounds[:, 0] = self.bounds[:, 0].clip(-1, 1)
+        self.bounds[:, 1] = self.bounds[:, 1].clip(0, 2)
+
+    def min_range(self, min_range: float):
+        low = self.bounds[:, 0] - self.bounds[:, 1]
+        high = self.bounds[:, 0] + self.bounds[:, 1]
+        low = low.clip(-1, 1)
+        high = high.clip(-1, 1)
+        diff = high - low
+        if min_range > 0:
+            invalid_indices = np.argwhere(diff < min_range)
+            self.bounds[invalid_indices, 1] += min_range / 2
+
+
+class MinPercentage(MatchingFunction):
+    """
+    A standard interval-based matching function producing multi-dimensional
+    hyperrectangular conditions. In effect, a lower bound (l) and
+    a distance proportion (p) are specified for each dimension.
+    An example x is matched iff
+    l_i <= x_i <= l_i + p_i * (max_i - l_i) for all dimensions i
+    """
+
+    def __init__(self, bounds: np.ndarray):
+        self.bounds = bounds
+
+    def __call__(self, X: np.ndarray):
+        lower = self.bounds[:, 0]
+        upper = lower + self.bounds[:, 1] * (1 - lower)
+        return np.all((lower <= X) & (X <= upper), axis=1)
+
+    @property
+    def volume_(self):
+        """Calculates the volume of the interval."""
+        lower = self.bounds[:, 0]
+        upper = lower + self.bounds[:, 1] * (1 - (-1))
+        diff = upper - lower
+        return np.prod(diff)
+
+    def copy(self):
+        return MinPercentage(self.bounds.copy())
+
+    def _validate_bounds(self, X: np.ndarray):
+        """Validates that bounds have the correct shape."""
+
+        if self.bounds.shape[1] != 2:
+            raise ValueError(f"specified bounds are not of shape (-1, 2), but {self.bounds.shape}")
+
+        if self.bounds.shape[0] != X.shape[1]:
+            raise ValueError(f"bounds- and input data dimension mismatch: {self.bounds.shape[0]} != {X.shape[1]}")
+
+    def clip(self, bounds: np.ndarray):
+        self.bounds[:, 0] = self.bounds[:, 0].clip(-1, 1)
+        self.bounds[:, 1] = self.bounds[:, 1].clip(0, 1)
+
+    def min_range(self, min_range: float):
+        lower = self.bounds[:, 0]
+        upper = lower + self.bounds[:, 1] * (1 - lower)
+        diff = upper - lower
+        if min_range > 0:
+            invalid_indices = np.argwhere(diff < min_range)
+            # Approximate increasing the width by min_range
+            self.bounds[invalid_indices, 0] -= min_range / 2
```

Comment:
Why do we do `/ 2` here but not in the next line?

Reply:
It approximates increasing the width by min_range, which is accomplished by moving the second bound by min_range and move the first by min_range / 2